packages/entity-server/src/routes/hierarchy/relations/RelationsResponseBuilder.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityServerModelRegistry } from '../../../types';
+import { EntityType } from '../../../models';
+import { formatEntitiesForResponse } from '../format';
+import { RelationsContext } from './types';
+
+type Pair = {
+  descendant: string;
+  ancestor: string;
+};
+
+export class RelationsResponseBuilder {
+  private readonly models: EntityServerModelRegistry;
+
+  private readonly ctx: RelationsContext & { ancestor: { type: string } };
+
+  private readonly groupBy: 'ancestor' | 'descendant';
+
+  constructor(
+    models: EntityServerModelRegistry,
+    ctx: RelationsContext,
+    groupBy: 'ancestor' | 'descendant' = 'ancestor',
+  ) {
+    this.models = models;
+    this.groupBy = groupBy;
+
+    const ancestorType = ctx.ancestor.type || ctx.entity.type;
+    if (ancestorType === null) {
+      throw new Error('No explicit ancestorType provided and entity type is null');
```

Comment:
Yeah, no, you're right. Usually with this kind of builder pattern you have no (or minimal) constructor params and then set properties before building. It's usually the `build()` function that does all the error checking before building. It can be very clean that way since you can have the various setters interacting with properties but it only does validation once.

That being said, it makes the typescript slightly more wordy to support that. You basically need to make the properties optional, and then validate that they're all present. I kinda wanted to rely on the constructor to be responsible for ensuring the ancestorType property is present so that the rest of the class can operate without having to check it's present all the time. I know it's not the cleanest OOP, but in this instance I think I prefer it 😄

Reply:
I understand, thanks for the explanation, certainly happy to leave this as is! 🙌